x-pack/plugins/encrypted_saved_objects/server/lib/encrypted_saved_objects_service.ts

```diff
@@ -257,33 +278,34 @@ export class EncryptedSavedObjectsService {
    * Generates string representation of the Additional Authenticated Data based on the specified saved
    * object type and attributes.
    * @param typeRegistration Saved object type registration parameters.
-   * @param id Id of the saved object.
+   * @param descriptor Descriptor of the saved object to get AAD for.
    * @param attributes All attributes of the saved object instance of the specified type.
    */
   private getAAD<T extends Record<string, unknown>>(
     typeRegistration: EncryptedSavedObjectTypeRegistration,
-    id: string,
+    descriptor: SavedObjectDescriptor,
     attributes: T
   ) {
     // Collect all attributes (both keys and values) that should contribute to AAD.
-    const attributesAAD = Array.from(Object.entries(attributes)).filter(([attributeKey]) => {
-      return (
+    const attributesAAD = {} as T;
+    for (const [attributeKey, attributeValue] of Object.entries(attributes)) {
+      if (
         !typeRegistration.attributesToEncrypt.has(attributeKey) &&
         (typeRegistration.attributesToExcludeFromAAD == null ||
           !typeRegistration.attributesToExcludeFromAAD.has(attributeKey))
-      );
-    });
+      ) {
+        attributesAAD[attributeKey] = attributeValue;
+      }
+    }
 
-    if (attributesAAD.length === 0) {
+    if (Object.keys(attributesAAD).length) {
       this.log.debug(
-        `The AAD for saved object "${typeRegistration.type}:${id}" does not include any attributes.`
+        `The AAD for saved object "${descriptorToArray(
+          descriptor
+        )}" does not include any attributes.`
       );
     }
 
-    // TODO: Use JSON.stringify instead of template string? Slower, but respects value types, e.g.
-    // Attributes: {prop1: 'prop1-value', prop2: 2}
-    // With template string: "prop1,prop1-value,prop2,2"
-    // With JSON.stringify:  "[[\"prop1\",\"prop1-value\"],[\"prop2\",2]]"
-    return `${typeRegistration.type},${id},${attributesAAD}`;
+    return stringify([...descriptorToArray(descriptor), attributesAAD]);
```

Comment:
Question: should we consider using `return stringify([descriptor, attributesAAD])` here instead? If we were to change the order returned by `descriptorToArray`, it'd break the original implementation. I'm not sure if this fear is relevant though.

Reply:
Do you have a some particular use case in mind? At the moment I can't think of any reason why we'd want to change the order of the elements returned by `descriptorToArray`. If we do this just to change the display order in other places where this method is used for some reason (e.g. logging)  then tests will break so that we know that we unintentionally tried to break AAD with this seemingly innocent change. If it wasn't the intention we can stop using `descriptorToArray` in `getAAD`. Moreover this order isn't something consumers can influence.

Having said I don't have a strong opinion on this, if it makes you feel more comfortable I can easily do that, just tell me :)

Reply:
>  then tests will break so that we know that we unintentionally tried to break AAD with this seemingly innocent change.

Fair enough, consider my concerns withdrawn.